### ISSUE TITLE
PDF flagging problems

### ISSUE THREAD

**Issue by raducoravu** [dita-ot/dita-ot#2459]
Problem 1, flagging does not work on DITA "step" elements:

Let's say I have a step in a DITA task:

```
        <step audience="abc">
            <cmd>Begin by cutting out all the dead branches.</cmd>
        </step>
```

and a DITAVAL file flagging it like:

```
 <val>
    <prop action="flag" att="audience" val="zuzu" color="yellow"/>
  </val>
```

The step in the PDF is not marked with yellow.
The problem seems to be generated by the fact that inside the XSLT stylesheet:

```
     plugins\org.dita.pdf2\xsl\fo\task-elements.xsl
```

the template:

```
   <xsl:template match="*[contains(@class, ' task/steps ')]/*[contains(@class, ' task/step ')]">
```

does not call the flagging template:

```
     <xsl:apply-templates select="*[contains(@class,' ditaot-d/ditaval-startprop ')]" mode="flag-attributes"/>
```

when it constructs the list-item.
But the problem is more general, none of the templates in the "task-elements.xsl" seem to call the flagging templates.

**Comment by raducoravu** [comment]
This mostly happens because not all templates call the "commonattributes" template which calls the flagging stuff.

**Comment by raducoravu** [comment]
As a side note the "flagging-preprocess_template.xsl" does not seem to be used anymore and should probably be removed.

**Comment by raducoravu** [comment]
Problem 2:

The flag image does not show up in the PDF. From what I understand the flag image should be given relative to the DITAVAL file.
So the DITAVAL is in the same folder with the DITA Map and it refers to the flagging image like this:

        <startflag imageref="common/figures/tool.png">

but in the *_MERGED.xml file the path to the image is not properly computed relative to the merged file.

**Comment by raducoravu** [comment]
Problem 3:

Flagging does not work on entire DITA topics and tasks, if the flagging attribute is set directly on the topic root element.
This is also reported here: https://github.com/dita-ot/dita-ot/issues/1545
The easiest fix I see for this would be that in the XSLT stylesheet:

     plugins\org.dita.pdf2\xsl\fo\commons.xsl

in the template:

          <xsl:template match="*" mode="commonTopicProcessing">

to wrap the apply-templates which generate the title and body:

             <xsl:apply-templates select="*[contains(@class, ' topic/title ')]"/>
            <xsl:apply-templates select="*[contains(@class, ' topic/prolog ')]"/>
            <xsl:apply-templates select="*[not(contains(@class, ' topic/title ')) and
                not(contains(@class, ' topic/prolog ')) and
                not(contains(@class, ' topic/topic '))]"/>

in a block which also calls the flagging processing:

        <fo:block>
            <xsl:apply-templates select="*[contains(@class,' ditaot-d/ditaval-startprop ')]" mode="flag-attributes"/>
            <xsl:apply-templates select="*[contains(@class, ' topic/title ')]"/>
            <xsl:apply-templates select="*[contains(@class, ' topic/prolog ')]"/>
            <xsl:apply-templates select="*[not(contains(@class, ' topic/title ')) and
                not(contains(@class, ' topic/prolog ')) and
                not(contains(@class, ' topic/topic '))]"/>
        </fo:block>

**Comment by drmacro** [comment]
The DITA Community PDF flagging plugin provides an override that fixes this. Don't know if that approach is appropriate here but it does fix the issue.

https://github.com/dita-community/org.dita-community.pdf.flagging

**Comment by raducoravu** [comment]
:+1: Another one of our users wanted flagging at topic level:
https://www.oxygenxml.com/forum/post39706.html#p39706

**Comment by robander** [comment]
Adding a zip that reproduces all of the problems listed here. It uses a flag image + color for the task, for the step, and (as a control) for the cmd. Also uses a topic from a sub-directory. Flags on the topic are ignored, flag image is used on the step but color is not, flag image is missing for topic that comes from subdir.
[flags.zip](https://github.com/dita-ot/dita-ot/files/974737/flags.zip)

**Comment by robander** [comment]
Problem 1 is an issue for: `step` (inside `steps` and `steps-unordered`), `substep`, `stepsection`, and `choice`. I've got a fix that handles all of these.

I'm actually not able to reproduce problem 3 anymore, at least with a simple map with tasks. The sample I attached earlier today was misleading, because the missing topic-level flags were also demonstrating problem 2 (topic was in a sub-dir).

**Comment by robander** [comment]
Problem 2 occurs because:

- DITA-OT currently assumes that DITAVAL and map are processed in the same directory; so a flag referenced as `myflag.jpg` will be in the same directory as the map.
- The flagging preprocess adjusts the image path when adding to topics in any other directory, so that it remains valid. But, it's inside of a `<foreign>` element, because this is DITAVAL markup (not valid DITA)
- The `topicmerge` process does not "un-adjust" the paths, so that a path adjusted to a value like `../myflag.jpg` is copied as-is, and is no longer valid in the map directory. The paths are left alone because this is foreign stuff and the merge doesn't touch foreign stuff.

There seem to be two straightforward ways to fix this; I've picked the second because it's easy and I think it should be usable for any other process that creates its own merged file:

1. Keep the current process that adjusts paths, but update topic merge so that it is aware of the `foreign` DITAVAL markup and reverses the adjust. I don't like this, it adds extra steps and exceptions for specific types of foreign content, and just feels wrong.
1. When adjusting the path, also store the original relative image path in a new attribute `dita-ot:original-imageref`. For PDF, where we know everything has been moved to the map directory, use that original relative path. Any other process that resolves everything in the map directory can also make use of the original reference.

**Comment by robander** [comment]
Er, nevermind about not reproducing problem #3. I get flag images, but realized I'm not getting the coloring.

**Comment by jelovirt** [comment]
Fixed by #2674

**Comment by raducoravu** [comment]
👍 I will also try to test this when I get back to work.
